### PR TITLE
QString replaced with mu::String in import Guitar Pro

### PR DESCRIPTION
**Pull request by DeeDonskih**
QString replaced with mu::String
Signed-off-by: Dmitry Donskikh <deedonskihdev@gmail.com>

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [ ] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp [hunk 1]
```diff
@@ -22,20 +22,20 @@ void GP67DomBuilder::buildGPDomModel(QDomElement* qdomElem)
...
-        { "MasterBars",    &masterBars },
-        { "Bars",                &bars },
-        { "Voices",        &voices },
-        { "Beats",               &beats },
-        { "Notes",               &notes },
-        { "Rhythms",       &rhythms },
-        { "GPVersion",   &gpversion },
-        { "Encoding",    &encoding }
+        { u"GPRevision",    &revision },
+        { u"Score",               &scoreNode },
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp:28]: I believe that it is possible to correct the alignment

Comment (DeeDonskih) [reply src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp:28]: Yep. I made it

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp [hunk 2]
```diff
@@ -864,19 +864,19 @@ void GPConverter::setUpGPScore(const GPScore* gpscore)
...
         Text* s = Factory::createText(_score->dummy(), TextStyleType::SUBTITLE);
-        QString str;
+        mu::String str;
         if (!gpscore->subTitle().isEmpty()) {
             str.append(gpscore->subTitle());
         }
         if (!gpscore->artist().isEmpty()) {
             if (!str.isEmpty()) {
-                str.append("\n");
+                str.append(mu::String("\n"));
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gpconverter.cpp:873]: needs just `u` and maybe needs add method `append(char16_t ch)` to `String`
(actualy, here conversion `const char* -> QString -> String`)

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp [hunk 3]
```diff
@@ -864,19 +864,19 @@ void GPConverter::setUpGPScore(const GPScore* gpscore)
...
             }
             str.append(gpscore->artist());
         }
         if (!gpscore->album().isEmpty()) {
             if (!str.isEmpty()) {
-                str.append("\n");
+                str.append(mu::String("\n"));
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gpconverter.cpp:879]: needs just `u` and maybe needs add method `append(char16_t ch)` to `String`
(actualy, here conversion `const char* -> QString -> String`)

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp [hunk 4]
```diff
@@ -1441,8 +1441,8 @@ void GPConverter::addFingering(const GPNote* gpnote, Note* note)
     }
 
     auto scoreFinger = [](const auto& str) {
-        if (str == "Open") {
-            return QString("O");
+        if (str == mu::String("Open")) {
+            return mu::String("O");
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gpconverter.cpp:1445]: needs just `u` 
(actualy, here conversion `const char* -> QString -> String`)

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp [hunk 5]
```diff
@@ -2256,9 +2256,9 @@ void GPConverter::addTimer(const GPBeat* beat, ChordRest* cr)
     int minutes = time / 60;
     int seconds = time % 60;
     Text* st = Factory::createText(cr->segment());
-    st->setPlainText(QString::number(minutes)
-                     + ":"
-                     + (seconds < 10 ? "0" + QString::number(seconds) : QString::number(seconds)));
+    st->setPlainText(mu::String::number(minutes)
+                     + mu::String(":")
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gpconverter.cpp:2260]: needs just `u` 
(actualy, here conversion `const char* -> QString -> String`)

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp [hunk 6]
```diff
@@ -2256,9 +2256,9 @@ void GPConverter::addTimer(const GPBeat* beat, ChordRest* cr)
...
+                     + (seconds < 10 ? mu::String("0") + mu::String::number(seconds) : mu::String::number(seconds)));
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gpconverter.cpp:2261]: needs just `u` 
(actualy, here conversion `const char* -> QString -> String`)

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp [hunk 7]
```diff
@@ -2655,15 +2655,16 @@ void GPConverter::clearDefectedGraceChord(ChordRestContainer& graceGhords)
...
+void GPConverter::addTextToNote(String string, Note* note)
 {
     Segment* segment = note->chord()->segment();
     StaffText* text = Factory::createStaffText(segment);
 
     if (!string.isEmpty()) {
         bool useHarmony = string[string.size() - 1] == '\\';
         if (useHarmony) {
-            string.resize(string.size() - 1);
+//            string.resize(string.size() - 1);
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gpconverter.cpp:2666]: needs remove comment

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 8]
```diff
@@ -677,8 +678,8 @@ void GuitarPro::addDynamic(Note* note, int d)
     if (!s->findAnnotation(ElementType::DYNAMIC, note->staffIdx() * VOICES, note->staffIdx() * VOICES + VOICES - 1)) {
         Dynamic* dyn = new Dynamic(s);
         // guitar pro only allows their users to go from ppp to fff
-        QString map_dyn[] = { "f", "ppp", "pp", "p", "mp", "mf", "f", "ff", "fff" };
-        dyn->setDynamicType(map_dyn[d]);
+        std::string map_dyn[] = { "f", "ppp", "pp", "p", "mp", "mf", "f", "ff", "fff" };
+        dyn->setDynamicType(mu::String::fromStdString(map_dyn[d]));
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:682]: better just add `u`, we can avoid conversion

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 9]
```diff
@@ -711,9 +712,9 @@ void GuitarPro::readVolta(GPVolta* gpVolta, Measure* m)
                 count++;
                 if (*iter == 1) {                 // we want this number to be displayed in the volta
                     if (voltaTextString == "") {
-                        voltaTextString += QString::number(count);
+                        voltaTextString += String::number(count);
                     } else {
-                        voltaTextString += "," + QString::number(count);
+                        voltaTextString += u"," + String::number(count);
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:717]: better replace to symbol (u`,`)

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 10]
```diff
@@ -730,9 +731,9 @@ void GuitarPro::readVolta(GPVolta* gpVolta, Measure* m)
                     // display all numbers in the volta from voltaSequence to the decimal
                     while (voltaSequence <= binaryNumber) {
                         if (voltaTextString == "") {
-                            voltaTextString = QString::number(voltaSequence);
+                            voltaTextString = String::number(voltaSequence);
                         } else {
-                            voltaTextString += "," + QString::number(voltaSequence);
+                            voltaTextString += u"," + String::number(voltaSequence);
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:736]: better replace to symbol (not string)

u`,`

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 11]
```diff
@@ -794,10 +795,10 @@ void GuitarPro::readLyrics()
     gpLyrics.fromBeat = readInt();
     gpLyrics.beatCounter = 0;
 
-    QString lyrics = readWordPascalString();
-    lyrics.replace(QRegularExpression("\n"), " ");
-    lyrics.replace(QRegularExpression("\r"), " ");
-    auto sl = lyrics.split(" ", Qt::KeepEmptyParts);
+    String lyrics = readWordPascalString();
+    lyrics.replace(std::regex("\n"), u" ");
+    lyrics.replace(std::regex("\r"), u" ");
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:800]: Why `regex` here?

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 12]
```diff
@@ -1509,18 +1510,18 @@ void GuitarPro::createSlur(bool hasSlur, staff_idx_t staffIdx, ChordRest* cr)
...
 //---------------------------------------------------------
 
-void GuitarPro::createOttava(bool hasOttava, int track, ChordRest* cr, QString value)
+void GuitarPro::createOttava(bool hasOttava, int track, ChordRest* cr, String value)
 {
     if (hasOttava && (ottava.at(track) == 0)) {
         Ottava* newOttava = new Ottava(score->dummy());
         newOttava->setTrack(track);
-        if (!value.compare("8va")) {
+        if (value=="8va") {
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:1518]: better add `u` (avoid conversion)

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 13]
```diff
@@ -1727,7 +1728,7 @@ bool GuitarPro2::read(IODevice* io)
             Segment* s = measure->getSegment(SegmentType::ChordRest, measure->tick());
             StaffText* st = new StaffText(s);
             //                  st->setTextStyleType(TextStyleType::STAFF);
-            st->setPlainText(QString("Capo. fret ") + QString::number(capo));
+            st->setPlainText(String("Capo. fret ") + String::number(capo));
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:1731]: better add `u` (avoid conversion)

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 14]
```diff
@@ -2443,7 +2444,7 @@ bool GuitarPro3::read(IODevice* io)
...
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:2447]: better add `u` (avoid conversion)

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 15]
```diff
@@ -2886,19 +2887,19 @@ static void addMetaInfo(MasterScore* score, GuitarPro* gp)
...
             Text* s = Factory::createText(m, TextStyleType::SUBTITLE);
-            QString str;
+            String str;
             if (!gp->subtitle.isEmpty()) {
                 str.append(gp->subtitle);
             }
             if (!gp->artist.isEmpty()) {
                 if (!str.isEmpty()) {
-                    str.append("\n");
+                    str.append(String("\n"));
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:2896]: needs just u and maybe needs add method append(char16_t ch) to String
(actualy, here conversion const char* -> QString -> String)

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 16]
```diff
@@ -2886,19 +2887,19 @@ static void addMetaInfo(MasterScore* score, GuitarPro* gp)
...
                 }
                 str.append(gp->artist);
             }
             if (!gp->album.isEmpty()) {
                 if (!str.isEmpty()) {
-                    str.append("\n");
+                    str.append(String("\n"));
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:2902]: needs just u and maybe needs add method append(char16_t ch) to String
(actualy, here conversion const char* -> QString -> String)

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 17]
```diff
@@ -3013,10 +3014,10 @@ static Score::FileError importScore(MasterScore* score, mu::io::IODevice* io)
         char ss[30];
         io->read((uint8_t*)(ss), 30);
         ss[l] = 0;
-        QString s(ss);
-        if (s.startsWith("FICHIER GUITAR PRO ")) {
+        String s(ss);
+        if (s.startsWith(String("FICHIER GUITAR PRO "))) {
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:3018]: better add u (avoid conversion)

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 18]
```diff
@@ -3013,10 +3014,10 @@ static Score::FileError importScore(MasterScore* score, mu::io::IODevice* io)
...
             s = s.mid(20);
-        } else if (s.startsWith("FICHIER GUITARE PRO ")) {
+        } else if (s.startsWith(String("FICHIER GUITARE PRO "))) {
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:3020]: better add u (avoid conversion)

File: src/importexport/guitarpro/internal/importgtp.h [hunk 19]
```diff
@@ -82,10 +82,12 @@ struct GPVolta {
 struct GPFermata {
     int index;
     int timeDivision;
-    QString type;
+//    QString type;
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.h:85]: needs remove comment

File: src/importexport/guitarpro/internal/importgtp.h [hunk 20]
```diff
@@ -82,10 +82,12 @@ struct GPVolta {
...
+    mu::String type;
 };
 
 struct GPLyrics {
+//    mu::StringList lyrics; //RESOLVE_QSTRR
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.h:90]: needs remove comment

File: src/importexport/guitarpro/internal/importgtp.h [hunk 21]
```diff
@@ -299,7 +311,7 @@ class GuitarPro
     GuitarPro(MasterScore*, int v);
     virtual ~GuitarPro();
     virtual bool read(mu::io::IODevice*) = 0;
-    QString error(GuitarProError n) const { return QString(errmsg[int(n)]); }
+    mu::String error(GuitarProError n) const { return mu::String(errmsg[int(n)]); }
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.h:314]: better use String::fromUtf8

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp (hunk 3 above)
Comment (DeeDonskih) [reply src/importexport/guitarpro/internal/gtp/gpconverter.cpp:879]: already made

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp [hunk 22]
```diff
@@ -1136,7 +1136,7 @@ void GPConverter::addContinuousSlideHammerOn()
                 auto midTick = (startTick + endTick) / 2;
                 Segment* segment = measure->getSegment(SegmentType::ChordRest, midTick);
                 StaffText* staffText = Factory::createStaffText(segment);
-                QString hammerText = (startNote->pitch() > endNote->pitch()) ? "P" : "H";
+                mu::String hammerText = (startNote->pitch() > endNote->pitch()) ? mu::String("P") : mu::String("H");
```
[the remark this answers is not in the document]
Comment (DeeDonskih) [reply src/importexport/guitarpro/internal/gtp/gpconverter.cpp:1139]: already made

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp (hunk 4 above)
Comment (DeeDonskih) [reply src/importexport/guitarpro/internal/gtp/gpconverter.cpp:1445]: already made

File: src/importexport/guitarpro/internal/importgtp.h [hunk 23]
```diff
@@ -239,11 +239,13 @@ class GuitarPro
     void read(void* p, qint64 len);
     int readUChar();
     int readChar();
-    QString readPascalString(int);
-    QString readWordPascalString();
-    QString readBytePascalString();
+//    QString readPascalString(int);
```
Comment (handrok) [inline src/importexport/guitarpro/internal/importgtp.h:242]: needs remove comment

File: src/framework/global/types/string.h [hunk 24]
```diff
@@ -428,6 +429,22 @@ class AsciiStringView
...
 inline String operator+(const char16_t* s1, const String& s2) { String t(s1); t += s2; return t; }
+
+// ============================
+// StringCodec
+// ============================
+
+class StringCodec
+{
+public:
+    static StringCodec* codecForName(const String& name);
```
Comment (handrok) [inline src/framework/global/types/string.h:440]: why not std::unique_ptr?

File: src/framework/global/types/string.h [hunk 25]
```diff
@@ -428,6 +429,22 @@ class AsciiStringView
 
 inline String operator+(char16_t s1, const String& s2) { String t(s1); t += s2; return t; }
 inline String operator+(const char16_t* s1, const String& s2) { String t(s1); t += s2; return t; }
+
+// ============================
+// StringCodec
+// ============================
+
+class StringCodec
```
Comment (handrok) [inline src/framework/global/types/string.h:437]: Do we need this class? We've already had similar functionality in mu::String class

File: src/framework/global/types/string.cpp [hunk 26]
```diff
@@ -1032,6 +1032,11 @@ double String::toDouble(bool* ok) const
     return toDouble_helper(ba.constChar(), ok);
 }
 
+float String::toFloat(bool *ok) const
+{
+    return static_cast<float>(toDouble(ok));
+}
```
Comment (igorkorsukov) [inline src/framework/global/types/string.cpp:1038]: It is better to update with the master, the master already has this code

File: src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp [hunk 27]
```diff
@@ -44,7 +44,7 @@ void GP67DomBuilder::buildGPDomModel(XmlDomElement* domElem)
         if (iter != nodeMap.end()) {
             *(iter->second) = node;
         } else {
-            QString nodeName = node.nodeName();
+            mu::String nodeName = node.nodeName();
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp:47]: `mu::` not need

File: src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp [hunk 28]
```diff
@@ -119,14 +119,14 @@ void GP67DomBuilder::buildGPMasterTracks(XmlDomNode* masterTrack)
...
+        if (nodeName == u"Automations") {
             masterTracks->setTempoMap(readTempoMap(&currentNode));
-        } else if (!nodeName.compare("RSE")) {
+        } else if (nodeName == u"RSE") {
             //! TODO volume and pan(balance) of mixer are set here
-        } else if (!nodeName.compare("Tracks")) {
+        } else if (nodeName == u"Tracks") {
             String tracks = currentNode.toElement().text();
-            size_t tracksCount = tracks.split(u' ').size();
+            size_t tracksCount = tracks.split(u" ").size();
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp:129]: better use symbol (char16_t), not string (const char16_t*) in this case

File: src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp [hunk 29]
```diff
@@ -308,27 +308,27 @@ std::unique_ptr<GPAudioTrack> GP67DomBuilder::createGPAudioTrack(XmlDomNode* met
 
 std::unique_ptr<GPMasterBar> GP67DomBuilder::createGPMasterBar(XmlDomNode* masterBarNode)
 {
-    static const std::set<QString> sUnused = {
-        "XProperties"
+    static const std::set<mu::String> sUnused = {
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gp67dombuilder.cpp:311]: `mu::` not need

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply src/importexport/guitarpro/internal/gtp/gpconverter.cpp:1139]: still bad code

File: src/importexport/guitarpro/internal/gtp/gpconverter.cpp [hunk 30]
```diff
@@ -1136,7 +1136,7 @@ void GPConverter::addContinuousSlideHammerOn()
                 auto midTick = (startTick + endTick) / 2;
                 Segment* segment = measure->getSegment(SegmentType::ChordRest, midTick);
                 StaffText* staffText = Factory::createStaffText(segment);
-                QString hammerText = (startNote->pitch() > endNote->pitch()) ? "P" : "H";
+                String hammerText = (startNote->pitch() > endNote->pitch()) ? String("P") : String("H");
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/gtp/gpconverter.cpp:1139]: still conversion from `const char*`

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 31]
```diff
@@ -1407,7 +1404,7 @@ void GuitarPro::setTempo(int temp, Measure* measure)
 
         TempoText* tt = new TempoText(segment);
         tt->setTempo(double(temp) / 60.0);
-        tt->setXmlText(QString("<sym>metNoteQuarterUp</sym> = %1").arg(temp));
+        tt->setXmlText(String("<sym>metNoteQuarterUp</sym> = %1").arg(temp));
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:1407]: needs add u

File: src/importexport/guitarpro/internal/importgtp.cpp [hunk 32]
```diff
@@ -125,8 +125,6 @@ GuitarPro::GuitarPro(MasterScore* s, int v)
 {
     score   = s;
     version = v;
-    std::string charset = configuration() ? configuration()->importGuitarProCharset() : "UTF-8";
-    _codec = QTextCodec::codecForName(QString::fromStdString(charset).toLatin1());
```
Comment (igorkorsukov) [inline src/importexport/guitarpro/internal/importgtp.cpp:129]: You need to remove the `_codec` field and the corresponding code that uses it

File: src/framework/global/types/string.h (hunk 25 above)
Comment (igorkorsukov) [reply src/framework/global/types/string.h:437]: Now this class is not needed

**Pull request merged**